crates/catalog/src/mutator.rs

```diff
@@ -30,12 +30,52 @@ impl CatalogMutator {
         self.client.as_ref()
     }
 
+    /// Commit the catalog state.
+    /// This persists the state to the metastore.
+    pub async fn commit_state(
+        &self,
+        catalog_version: u64,
+        state: CatalogState,
+    ) -> Result<Arc<CatalogState>> {
+        let client = match &self.client {
+            Some(client) => client,
+            None => return Err(CatalogError::new("metastore client not configured")),
+        };
+
+        let state = match client.commit_state(catalog_version, state).await {
+            Ok(state) => state,
+            Err(CatalogError {
+                msg,
+                strategy: Some(ResolveErrorStrategy::FetchCatalogAndRetry),
+            }) => {
+                // Go ahead and refetch the catalog and retry the mutation.
+                //
+                // Note that this relies on metastore _always_ being stricter
+                // when validating mutations. What this means is that retrying
+                // here should be semantically equivalent to manually refreshing
+                // the catalog and rerunning and replanning the query.
```

Comment:
but only once? if this fails do we bail out and retry at a higher level anyway? Why not implement this logic in the commit_state method rather than here? is there a limit to the depth of retries that we should allow, or does this just wait to hit the stack or recursion limit?

The debug statement should also, probably, include slightly more information: which deployment/database-id/catalog as well as the call depth/number of retries, or version number, if possible.

Reply:
Honestly, i don't have a super good understanding of this code, it was just copied from the original function.

Reply:
In the original `mutate` method, this retry exists mostly for cloud + HA metastores. A glaredb instance might have an out-of-date catalog if some other instance previously made a change and we haven't updated our cached state yet. The comment is just saying that we retry the mutation against the now up-to-date version, and would (should) semantically be the same as the user running the query again.

This is only retried once since catalog changes are infrequent, and the likelihood that a separate change is made between when we get the updated state back and when we try to apply the mutations again is low.

Reply:
I actually think that it's retried infinitely because of the recursive call.

Reply:
> I actually think that it's retried infinitely because of the recursive call.

it's not calling `self.commit_state` but instead `client.commit_state` so this isn't recursive